cartographer/mapping/internal/pose_graph/optimization_problem_interface.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_MAPPING_INTERNAL_POSE_GRAPH_OPTIMIZATION_PROBLEM_INTERFACE_H_
+#define CARTOGRAPHER_MAPPING_INTERNAL_POSE_GRAPH_OPTIMIZATION_PROBLEM_INTERFACE_H_
+
+#include <map>
+#include <set>
+#include <vector>
+
+#include "Eigen/Core"
+#include "Eigen/Geometry"
+#include "cartographer/common/time.h"
+#include "cartographer/mapping/id.h"
+#include "cartographer/mapping/pose_graph_interface.h"
+#include "cartographer/sensor/fixed_frame_pose_data.h"
+#include "cartographer/sensor/imu_data.h"
+#include "cartographer/sensor/map_by_time.h"
+#include "cartographer/sensor/odometry_data.h"
+
+namespace cartographer {
+namespace mapping {
+namespace pose_graph {
+
+// Implements the SPA loop closure method.
+template <typename NodeDataType, typename SubmapDataType,
+          typename RigidTransformType>
+class OptimizationProblemInterface {
+ public:
+  using Constraint = PoseGraphInterface::Constraint;
+  using LandmarkNode = PoseGraphInterface::LandmarkNode;
+  using NodeData = NodeDataType;
+  using SubmapData = SubmapDataType;
+  using RigidTransform = RigidTransformType;
```

Comment:
Please, remove these typedefs and define them in .cc files only. It won't change a lot here.

Reply:
Done